frame/balances/src/lib.rs

```diff
@@ -1173,12 +1227,31 @@ impl<T: Config<I>, I: Instance> ReservableCurrency<T::AccountId> for Module<T, I
 	) -> (Self::NegativeImbalance, Self::Balance) {
 		if value.is_zero() { return (NegativeImbalance::zero(), Zero::zero()) }
 
-		Self::mutate_account(who, |account| {
-			// underflow should never happen, but it if does, there's nothing to be done here.
-			let actual = cmp::min(account.reserved, value);
-			account.reserved -= actual;
-			(NegativeImbalance::new(actual), value - actual)
-		})
+		// NOTE: `mutate_account` may fail if it attempts to reduce the balance to the point that an
+		//   account is attempted to be illegally destroyed.
+
+		for attempt in 0..2 {
+			match Self::mutate_account(who, |account| {
+				let best_value = match attempt {
+					0 => value,
+					// If acting as a critical provider (i.e. first attempt failed), then ensure
+					// slash leaves at least the ED.
+					_ => value.min(account.free + account.reserved - T::ExistentialDeposit::get()),
+				};
+
+				let actual = cmp::min(account.reserved, best_value);
+				account.reserved -= actual;
+
+				// underflow should never happen, but it if does, there's nothing to be done here.
+				(NegativeImbalance::new(actual), value - actual)
+			}) {
+				Ok(r) => return r,
+				Err(_) => (),
+			}
+		}
```

Comment:
For what it's worth, I don't like how tight this coupling is while being so indirect. (We assume a certain error lead to the operation failing the first time, without looking at the error at all, and then attempt to fix it.)

Reply:
The way I view this is kind of "best effort" loop.

First we try to slash full amount, best effort. Oops that didn't work, now we try to slash up to ED amount, cause we know about one obvious fail case. Oops that didn't work, return anyway, we tried our best.

Reply:
:ok_hand: